### PR TITLE
Plugin idp opennebula

### PR DESCRIPTION
**Pull request by ANKerD**

### CODE REVIEW

File: src/test/java/cloud/fogbow/common/util/SystemUserUtilTest.java [hunk 1]
```diff
@@ -20,7 +25,30 @@ public void testSerializationDeserealizationOfSystemUserObject() throws Unexpect
         Assert.assertEquals(systemUser, recoveredSystemUser);
     }
 
+    // Test if size of content is going to overflow the limit a SystemUSer is allowed
```
Comment (wassoares) [inline src/test/java/cloud/fogbow/common/util/SystemUserUtilTest.java]: Standardize the test by following the example:
    // test case: ...
    @Test(expected = Exception.class) // verify
    public void test() {
        // set up
        // exercise
    }

File: src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+        PowerMockito.mockStatic(OpenNebulaClientUtil.class);
+        BDDMockito.given(OpenNebulaClientUtil.createClient(Mockito.anyString(), Mockito.anyString())).willReturn(client);
+    }
+
+    @Test
+    public void testAuthenticateSuccess() throws FogbowException {
+        // setup
+        Mockito.doReturn(true).when(this.plugin).isAuthenticated(Mockito.anyString());
+        HashMap<String,String> credentials = getCredentials(FAKE_USER_ID, FAKE_USER_PASSWORD);
```
Comment (thiagomanel) [inline src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java]: why are you not using the Map interface here?

File: src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        OpenNebulaUser openNebulaUser = this.plugin.getCloudUser(credentials);
+
+        // verify
+        Assert.assertNotNull(openNebulaUser);
+        Assert.assertEquals(FAKE_USER_ID, openNebulaUser.getId());
+        Mockito.verify(this.plugin).isAuthenticated(Mockito.anyString());
+    }
+
+    @Test(expected = UnauthenticatedUserException.class)
+    public void testAuthenticateUnsuccess() throws FogbowException {
```
Comment (thiagomanel) [inline src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java]: I think there is no such word "Unsuccess"

File: src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        Mockito.doReturn(false).when(this.plugin).isAuthenticated(Mockito.anyString());
+        HashMap<String,String> credentials = getCredentials(FAKE_USER_ID, INVALID_PASSWORD);
+
+        // exercise
```
Comment (thiagomanel) [inline src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java]: remove all unnecessary `OpenNebulaUser openNebulaUser =` declarations

File: src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    @Test(expected = InvalidParameterException.class)
+    public void testAuthenticateWithEmptyParams() throws FogbowException {
+        HashMap<String,String> credentials = getCredentials(FAKE_USER_ID, "");
+
+        // exercise
+        OpenNebulaUser openNebulaUser = this.plugin.getCloudUser(credentials);
+    }
+
+    @Test
+    public void testIsAuthenticatedSuccessWithoutMockingHelperMethod() throws FogbowException {
```
Comment (thiagomanel) [inline src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java]: why do we need this testcase?

File: src/main/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPlugin.java [hunk 6]
```diff
@@ -42,7 +42,8 @@ public OpenNebulaUser getCloudUser(Map<String, String> userCredentials) throws F
         String username = userCredentials.get(OpenNebulaConstants.USERNAME);
         String password = userCredentials.get(OpenNebulaConstants.PASSWORD);
 
-        if (username == null || password == null) {
+        if (username == null || username.trim().isEmpty()
+                || password == null || password.trim().isEmpty()) {
```
Comment (thiagomanel) [inline src/main/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPlugin.java]: do we need final declarations in line 64:65?

File: src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    public void setup() throws UnexpectedException {
+        this.plugin = Mockito.spy(new OpenNebulaIdentityProviderPlugin(FAKE_IDENTITY_URL));
+
+        Client client = Mockito.mock(Client.class);
+
+        PowerMockito.mockStatic(OpenNebulaClientUtil.class);
+        BDDMockito.given(OpenNebulaClientUtil.createClient(Mockito.anyString(), Mockito.anyString())).willReturn(client);
+    }
+
+    @Test
```
[the remark this answers is not in the document]
Comment (ANKerD) [reply src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java]: Add test descriptions.

File: src/test/java/cloud/fogbow/common/util/PropertiesUtilTest.java [hunk 8]
```diff
@@ -12,9 +17,33 @@
     private static final String FAKE_PROPERTIES_FILE_NAME = "fake-properties.conf";
     private static final String NON_EXISTENT_FILE_NAME = "non-existent-file-name.conf";
 
+    // test if the loadProperties returns the expected properties when a
```
[the remark this answers is not in the document]
Comment (ANKerD) [reply src/test/java/cloud/fogbow/common/util/PropertiesUtilTest.java]: Changed.

File: src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java (hunk 5 above)
Comment (ANKerD) [reply src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java]: To make sure the helper method behaves appropriately.

File: src/test/java/cloud/fogbow/common/util/PropertiesUtilTest.java [hunk 9]
```diff
@@ -12,9 +17,33 @@
...
+        Assert.assertEquals(props.getProperty("fake_property1"), "f1");
+        Assert.assertEquals(props.getProperty("fake_property2"), "f2");
+        Assert.assertEquals(props.getProperty("fake_property3"), "f3");
+    }
+
+    // test if a FatalErrorException is throwed when a non existent filepath is passed to loadProperties
+    @Test(expected = FatalErrorException.class) // verify
+    public void testLoadPropertiesWithANonExistentFile() {
+        // setup // exercise
+        Properties props = PropertiesUtil.loadProperties("non-existent-file");
```
[the remark this answers is not in the document]
Comment (ANKerD) [reply src/test/java/cloud/fogbow/common/util/PropertiesUtilTest.java]: Done.

File: src/test/java/cloud/fogbow/common/util/PropertiesUtilTest.java [hunk 10]
```diff
@@ -32,6 +61,32 @@ public void testGetProperties() {
         }
     }
 
+    @Test
```
[the remark this answers is not in the document]
Comment (ANKerD) [reply src/test/java/cloud/fogbow/common/util/PropertiesUtilTest.java]: Solved.

File: src/test/java/cloud/fogbow/common/util/PropertiesUtilTest.java [hunk 11]
```diff
@@ -32,6 +61,32 @@ public void testGetProperties() {
...
+
+        Properties fakeProperties = PropertiesUtil.readProperties(fakePropertiesFileList);
+
+        // exercise/verify
+
+        for (int i = 1; i <= 6; i++){
+            String expectedFakePropertyKey = "fake_property" + i;
+            String expectedFakePropertyValue = "f" + i;
+
+            String fakeProperty = fakeProperties.getProperty(expectedFakePropertyKey);
```
[the remark this answers is not in the document]
Comment (ANKerD) [reply src/test/java/cloud/fogbow/common/util/PropertiesUtilTest.java]: Done in other PR.

File: src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java (hunk 2 above)
Comment (ANKerD) [reply src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java]: Done.

File: src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java (hunk 3 above)
Comment (ANKerD) [reply src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java]: Fixed.

File: src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java (hunk 4 above)
Comment (ANKerD) [reply src/test/java/cloud/fogbow/common/plugins/cloudidp/opennebula/OpenNebulaIdentityProviderPluginTest.java]: Solved.

**Pull request merged**